pkg/skaffold/build/deps.go

```diff
@@ -0,0 +1,114 @@
+package build
+
+import (
+	"path/filepath"
+	"sort"
+	"strings"
+
+	"github.com/GoogleCloudPlatform/skaffold/pkg/skaffold/config"
+	"github.com/pkg/errors"
+	"github.com/sirupsen/logrus"
+)
+
+type DependencyMap struct {
+	artifacts       []*config.Artifact
+	pathToArtifacts map[string][]*config.Artifact
+}
+
+type DependencyResolver interface {
+	GetDependencies(a *config.Artifact) ([]string, error)
+}
+
+//TODO(@r2d4): Figure out best UX to support configuring this blacklist
+var ignoredPrefixes = []string{"vendor", ".git"}
+
+func (d *DependencyMap) Paths() []string {
+	allPaths := []string{}
+	for path := range d.pathToArtifacts {
+		allPaths = append(allPaths, path)
+	}
+	sort.Strings(allPaths)
+	return allPaths
+}
+
+func (d *DependencyMap) ArtifactsForPaths(paths []string) []*config.Artifact {
+	m := map[*config.Artifact]struct{}{}
+	for _, p := range paths {
+		artifacts := d.pathToArtifacts[p]
+		for _, a := range artifacts {
+			m[a] = struct{}{}
+		}
+	}
+	artifacts := []*config.Artifact{}
+	for a := range m {
+		artifacts = append(artifacts, a)
+	}
+	return artifacts
+}
+
+func NewDependencyMap(artifacts []*config.Artifact, res DependencyResolver) (*DependencyMap, error) {
+	m, err := pathToArtifactMap(artifacts, res)
+	if err != nil {
+		return nil, errors.Wrap(err, "generating path to artifact map")
+	}
+	return &DependencyMap{
+		artifacts:       artifacts,
+		pathToArtifacts: m,
+	}, nil
+}
+
+// path must be an absolute path
+func isIgnored(workspace, path string) (bool, error) {
```

Comment:
Have you considered making a type for AbsolutePath, with conversion functions? That could get us an element of type-safety and catch a bunch of potential bugs.

I tried this out somewhere else (I forget where), but it seemed to work pretty well.